### DISCUSSION TITLE
Is the new version of libmesh slower than the old version?

### DISCUSSION THREAD

**Discussion by LHappyCureall** [libMesh/libmesh#3679] [General]
Now I use the libmesh-1.8.0-pre petsc-3.19. When I run system_of_equations ex2  :` mpirun -n 4 ./example-opt`, it is much much slower than the libmesh-0.9.3 petsc-3.4 version in the same example. ( I've modified them to the same conditions; only the syntax is different.)

I've always been convinced that the new version of libmesh and petsc is better, but why is it running so slow?  

Here is my configure :  `./configure --prefix=/home/lhc/petsc3.19/libmesh/opt2  PETSC_ARCH=test2 PETSC_DIR=/home/lhc/petsc3.19/petsc-3.19.6 CXXFLAGS='-O3 -fPIC' CFLAGS=-O3 FFLAGS=-O3 F77=gfortran --enable-fPIC --enable-parmesh --enable-blocked-storage --enable-perflog  --with-method=opt --enable-mpi --enable-petsc  --with-dof-id-bytes=8 --disable-fparser --enable-slepc --enable-amr --enable-static=yes`

I also want to know how to speed up the operation efficiency, thank you!

**Comment by roystgnr** [comment]
We've never noticed any such regression, but in the decade between the libMesh 0.9.3 release and now CPU cores have gotten at least twice as fast and four times as plentiful, so maybe some gradual regression got masked?

Is there any way to reproduce any changes on a modern system?  I can't even build PETSc 3.4 right now because its configure doesn't work with python 3 and python 2 isn't available on Ubuntu 23.10.

Can you post the perflog differences?  If there's any regression, I can't reproduce it, even just looking at modern libMesh: if I run systems_of_equations_ex2 on 4 cores, I see:

```
| PetscLinearSolver                                                                                                   |
|   solve()                              47         0.5246      0.011161    0.5246      0.011161    77.57    77.57    |
```

The solve() is mostly just a shim around a PETSc KSP there.  libMesh could execute infinity percent faster and that example would only finish 22.5% sooner (in ~.5 seconds instead of ~.7?), which puts a pretty tight upper limit on the possible magnitude of any regression there.

Is it possible you accidentally built your PETSc 3.19 in debug mode?  I think at some point they switched that to be the default, with optimized builds requiring `--with-debugging=no` and optimization flags to be manually added to their configure.

**Reply by LHappyCureall** [reply]
Thank you very much, Roy. 

I am sure in a` '--with-debugging=no'` mode, here is my PETSc-3.19 configure: 
```
'Configure Options: --configModules=PETSc.Configure --optionsModule=config.compilerOptions CXXFLAGS=-fPIC --with-mpi-dir=/usr --download-f2cblaslapack=1 --with-fc=0 --with-debugging=0 COPTFLAGS=-O3 CXXOPTFLAGS=-O3 --download-hypre=1 --with-c2html=0 --with-shared-libraries=0 --with-64-bit-indices
Working directory: /home/lhc/petsc19/petsc-3.19.1`
```
<br>

> Can you post the perflog differences?

And I have compared the perflog  `PetscLinearSolver `:  (same program code) libMesh-1.8.0-pre  runs faster than the old version, XD. 

```
|   solve()        - libMesh-0.9.3                 225        41.5782     0.184792    41.5782     0.184792    98.09    98.09    |
|   solve()        - libMesh-1.8.0-pre           225        8.8405      0.039291    8.8405      0.039291    93.95    93.95    |
```

I need to do a few more comparisons to determine the advantages of the new libMesh-1.8.0-pre, LOL.
<br>

>  I can't even build PETSc 3.4 right now because its configure doesn't work with python 3 and python 2 isn't available on Ubuntu 23.10.

Firstly, you may need to lower the GCC version to 7.5 in Ubuntu 23 manually.   Then, install python2 `sudo apt install python2`.  Finally, switch python2 to the default version `sudo update-alternatives --install /usr/bin/python python /usr/bin/python2 1`.

But I know a more convenient way: Ubuntu 18. LTS(which I'm using right now), both python2 and python3 were preinstalled.  

I appreciate the help you’ve given me!
